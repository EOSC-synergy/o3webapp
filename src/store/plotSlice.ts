import { createSlice } from '@reduxjs/toolkit';
import type { Latitude } from 'utils/constants';
import {
    ALL_REGIONS_ORDERED,
    END_YEAR,
    latitudeBands,
    O3AS_PLOTS,
    START_YEAR,
} from 'utils/constants';
import { HYDRATE } from 'next-redux-wrapper';

/**
 * These months are selected when the webapp starts
 *
 * @category PlotSlice
 */
const DEFAULT_MONTHS: number[] = [1, 2, 12];

// used for return
export type RegionBasedXRange = { regions: number[] };
// used for zm
export type YearsBasedXRange = {
    years: {
        minX: number;
        maxX: number;
    };
};
export type XRange = RegionBasedXRange | YearsBasedXRange;
export type YRange = {
    minY: number;
    maxY: number;
};

export type PlotSpecific = {
    // initial settings for tco3_zm
    [O3AS_PLOTS.tco3_zm]: {
        // the title shown in the apexcharts generated chart
        title: string;
        displayXRange: YearsBasedXRange;
        // gets adjusted automatically on each request
        displayYRange: YRange;
    };
    // initial settings for tco3_return
    [O3AS_PLOTS.tco3_return]: {
        title: string;
        displayXRange: RegionBasedXRange;
        // gets adjusted automatically on each request
        displayYRange: YRange;
        userRegionName?: string;
    };
};

type PlotSpecifics<
    Id extends O3AS_PLOTS,
    P extends keyof PlotSpecific[O3AS_PLOTS]
> = Id extends O3AS_PLOTS
    ? {
          plotId: Id;
      } & {
          [key in P]: PlotSpecific[Id][key];
      }
    : never;
type PickPlotSpecific<P extends keyof PlotSpecific[O3AS_PLOTS]> = PlotSpecifics<O3AS_PLOTS, P>;

export type PlotState = {
    plotId: O3AS_PLOTS;
    generalSettings: {
        location: Latitude;
        months: number[];
    };
    plotSpecificSettings: PlotSpecific;
};

export type GlobalPlotState = {
    plot: PlotState;
};

/**
 * The initial state of the plotSlice defines the data structure in the store. Each plot has its own
 * settings i.e. title, location etc.
 *
 * IF you change this initial state you have to adapt the first test in the corresponding test file,
 * that tests the initial state.
 */
export const initialState: PlotState = {
    plotId: O3AS_PLOTS.tco3_zm, // the initially active plot
    // maps plot ids to their settings
    generalSettings: {
        location: latitudeBands[6].value, // Global Region
        months: DEFAULT_MONTHS,
    },

    plotSpecificSettings: {
        // initall settings for tco3_zm
        [O3AS_PLOTS.tco3_zm]: {
            title: 'OCTS Plot', // the title shown in the apexcharts generated chart
            displayXRange: {
                years: {
                    minX: START_YEAR,
                    maxX: END_YEAR,
                },
            },
            displayYRange: { minY: 0, maxY: 0 }, // gets adjusted automatically on each request
        },
        // initall settings for tco3_return
        [O3AS_PLOTS.tco3_return]: {
            title: 'Return/Recovery Plot',
            displayXRange: {
                regions: [...Array(ALL_REGIONS_ORDERED.length).keys()], // implicitly refers to ALL_REGIONS_ORDERED by storing only the indices
            },
            displayYRange: { minY: 0, maxY: 0 }, // gets adjusted automatically on each request
        },
    },
};

type Payload<T> = {
    payload: T;
};

/**
 * The plotSlice is generated by the redux toolkit. The reducers are defined here and the
 * corresponding actions are auto-generated.
 *
 * @category PlotSlice
 */
const plotSlice = createSlice({
    name: 'plot',
    initialState,
    reducers: {
        /**
         * This reducer accepts an action object returned from setActivePlotId() and calculates the
         * new state based on the action and the action data given in action.payload.
         *
         * In this case the active plotId is set to the given string.
         *
         * @example
         *     dispatch(setActivePlotId({ id: 'tco3_zm' }));
         *     s;
         *
         * @param state The current store state of: state/plot
         * @param plotId A string that contains the new plot id
         */
        setActivePlotId(
            state: PlotState,
            { payload: { plotId } }: Payload<{ plotId: O3AS_PLOTS }>
        ) {
            state.plotId = plotId;
            console.log('active plot', plotId);
        },

        setTitle(state: PlotState, { payload: { title } }: Payload<{ title: string }>) {
            state.plotSpecificSettings[state.plotId].title = title;
        },

        /**
         * This reducer accepts an action object returned from setLocation() and calculates the new
         * state based on the action and the action data given in action.payload.
         *
         * In this case the current plot location is set with the provided new latitude values.
         *
         * @param state The current store state of: state/plot
         * @param payload
         * @param payload.minLat A number specifying the minimum latitude
         * @param payload.maxLat A number specifying the maximum latitude
         * @exmaple dispatch(setLocation({minLat: -90, maxLat: 90}));
         */
        setLocation(state: PlotState, { payload }: Payload<Latitude>) {
            state.generalSettings.location = payload;
        },

        /**
         * This reducer accepts an action object returned from setDisplayXRange() and calculates the
         * new state based on the action and the action data given in action.payload.
         *
         * For the tco3_zm the current displayXRange is updated with the given min and max x values,
         * i.e. you have to pass {years: {minX: 1960, maxX: 2100}} as payload.
         *
         * For the tco3_return the current displayXRange is updated with the given regions values,
         * i.e. you have to pass {regions: [0, 1, 2]} as payload.
         *
         * @example
         *     dispatch(setDisplayXRange({ years: { minX: 1960, maxX: 2100 } }));
         *
         * @example
         *     dispatch(setDisplayXRange({ regions: [0, 1, 2] }));
         *
         * @param state The current store state of: state/plot
         * @param payload
         * @param payload.years.minX A number specifying the start of the x range for tco3_zm
         * @param payload.years.maxX A number specifying the end of the x range for tco3_zm
         * @param payload.regions An array specifying the selected regions (on the x-axis) for the
         *   tco3_return
         */
        setDisplayXRangeForPlot(
            state: PlotState,
            { payload }: { payload: PickPlotSpecific<'displayXRange'> }
        ) {
            if (payload.plotId === 'tco3_zm') {
                const {
                    years: { minX, maxX },
                } = payload.displayXRange;
                const xRange = state.plotSpecificSettings[payload.plotId].displayXRange;
                xRange.years.minX = minX;
                xRange.years.maxX = maxX;
            } else if (payload.plotId === O3AS_PLOTS.tco3_return) {
                const { regions } = payload.displayXRange;
                state.plotSpecificSettings[payload.plotId].displayXRange.regions = regions;
            }
        },

        /**
         * This reducer accepts an action object returned from setDisplayYRange() and calculates the
         * new state based on the action and the action data given in action.payload.
         *
         * In this case the current displayYRange is updated with the given min and max y values.
         * "Current" is determined by the current active plotId.
         *
         * @example
         *     dispatch(setDisplayYRange({ minY: 200, maxY: 400 }));
         *
         * @param state The current store state of: state/plot
         * @param payload
         * @param payload.minY A number specifying the start of the y range
         * @param payload.maxY A number specifying the end of the y range
         */
        setDisplayYRange(state: PlotState, { payload }: { payload: YRange }) {
            if (payload.minY === null || payload.maxY === null) {
                return;
            }
            if (!Number.isFinite(payload.minY) || !Number.isFinite(payload.maxY)) {
                return;
            }
            if (isNaN(payload.minY) || isNaN(payload.maxY)) {
                return;
            }

            state.plotSpecificSettings[state.plotId].displayYRange = payload;
        },

        /**
         * This reducer accepts an action object returned from setDisplayYRange() and calculates the
         * new state based on the action and the action data given in action.payload.
         *
         * In this case the displayYRange for the given plotId is updated with the given min and max
         * y values.
         *
         * @example
         *     dispatch(setDisplayYRange({ minY: 200, maxY: 400 }));
         *
         * @param state The current store state of: state/plot
         * @param plotId Tco3_zm | tco3_return
         * @param payload.plotId
         * @param payload.minY A number specifying the start of the y range
         * @param payload.maxY A number specifying the end of the y range
         */
        setDisplayYRangeForPlot(
            state: PlotState,
            { payload }: { payload: PickPlotSpecific<'displayYRange'> }
        ) {
            const {
                plotId,
                displayYRange: { minY, maxY },
            } = payload;
            if (minY === null || maxY === null) {
                return;
            }
            if (!Number.isFinite(minY) || !Number.isFinite(maxY)) {
                return;
            }
            if (isNaN(minY) || isNaN(maxY)) {
                return;
            }

            const displayYRange = state.plotSpecificSettings[plotId].displayYRange;
            displayYRange.minY = minY;
            displayYRange.maxY = maxY;
        },

        /**
         * This reducer accepts an action object returned from setMonths() and calculates the new
         * state based on the action and the action data given in action.payload.
         *
         * In this case the current selected months are updated to the given array.
         *
         * @example
         *     dispatch(setMonths({ months: [3, 4, 5] }));
         *
         * @param state The current store state of: state/plot
         * @param payload The payload is an object containing the given data
         * @param payload.months An array of integers describing the new months
         */
        setMonths(state: PlotState, { payload }: { payload: { months: number[] } }) {
            state.generalSettings.months = payload.months;
        },
    },
    extraReducers: {
        [HYDRATE]: (state, action) => {
            return {
                ...state,
                ...action.payload.subject,
            };
        },
    },
});

/**
 * The here listed actions are exported and serve as an interface for the view (our React
 * components).
 */
export const {
    setActivePlotId,
    setTitle,
    setLocation,
    setDisplayXRangeForPlot,
    setDisplayYRange,
    setDisplayYRangeForPlot,
    setMonths,
} = plotSlice.actions;

/**
 * The reducer combining all reducers defined in the plot slice. This has to be included in the
 * redux store, otherwise dispatching the above defined actions wouldn't trigger state updates.
 */
export default plotSlice.reducer;

/**
 * This selector allows components to select the current plot id from the store. The plot id is a
 * string using the same naming as the o3as api e.g. tco3_zm or tco3_return
 *
 * @category PlotSlice
 * @function
 * @param state The global redux state
 * @returns The current active plot id
 */
export const selectPlotId = (state: GlobalPlotState) => state.plot.plotId;

/**
 * This selector allows components to select the current plot title from the store.
 *
 * @category PlotSlice
 * @function
 * @param state The global redux state
 * @returns The current active plot title
 */
export const selectPlotTitle = (state: GlobalPlotState) =>
    state.plot.plotSpecificSettings[state.plot.plotId].title;

/**
 * This selector allows components to select the current plot location from the store. The location
 * is an object containing a minLat and maxLat attribute.
 *
 * @category PlotSlice
 * @function
 * @param state The global redux state
 * @returns Holds the current location that includes a minLat and maxLat attribute.
 */
export const selectPlotLocation = (state: GlobalPlotState) => state.plot.generalSettings.location;

/**
 * This selector allows components to select the current x range from the store.
 *
 * @category PlotSlice
 * @function
 * @param plotId Plot type (needed for typing)
 * @returns Holds the current x range that includes minX and maxX
 */
export const selectPlotXRange =
    <T extends O3AS_PLOTS>(plotId: T) =>
    (state: GlobalPlotState): PlotSpecific[T]['displayXRange'] =>
        state.plot.plotSpecificSettings[plotId].displayXRange;

/**
 * This selector allows components to select the current y range from the store.
 *
 * @category PlotSlice
 * @function
 * @param state The global redux state
 * @returns Holds the current x range that includes minY and maxY
 */
export const selectPlotYRange = (state: GlobalPlotState) =>
    state.plot.plotSpecificSettings[state.plot.plotId].displayYRange;

/**
 * This selector allows components to select the current selected months from the store.
 *
 * @category PlotSlice
 * @function
 * @param state The global redux state
 * @returns Array of integers describing the current selected months
 */
export const selectPlotMonths = (state: GlobalPlotState) => state.plot.generalSettings.months;

/**
 * This selector allows components to select the current user region name.
 *
 * @function
 * @param state The global redux state
 * @returns The user region name
 */
export const selectUserRegionName = (state: GlobalPlotState) =>
    state.plot.plotSpecificSettings.tco3_return.userRegionName;
