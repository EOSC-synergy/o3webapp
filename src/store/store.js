import { applyMiddleware, combineReducers, configureStore, createStore } from '@reduxjs/toolkit';
import plotReducer from './plotSlice/plotSlice';
import modelsReducer, { setModelsOfModelGroup } from './modelsSlice/modelsSlice';
import referenceReducer from './referenceSlice/referenceSlice';
import apiReducer from '../services/API/apiSlice/apiSlice';
import { DEFAULT_MODEL_GROUP } from '../utils/constants';
import { HYDRATE, createWrapper } from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk';

const bindMiddleware = (middleware) => {
    if (process.env.NODE_ENV !== 'production') {
        const { composeWithDevTools } = require('redux-devtools-extension');
        return composeWithDevTools(applyMiddleware(...middleware));
    }
    return applyMiddleware(...middleware);
};

const combinedReducer = combineReducers({
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
    api: apiReducer,
});

const reducer = (state, action) => {
    if (action.type === HYDRATE) {
        const nextState = {
            ...state, // use previous state
            ...action.payload, // apply delta from hydration
        };
        // if (state.count.count) nextState.count.count = state.count.count; // preserve count value on client side navigation
        return nextState;
    } else {
        return combinedReducer(state, action);
    }
};

let store;

const initStore = () => {
    store = createStore(reducer, bindMiddleware([thunkMiddleware]));
    return store;
};

export { store };

export const wrapper = createWrapper(initStore);

export function createTestStore() {
    const store = configureStore({
        reducer: combinedReducer,
    });
    store.dispatch(setModelsOfModelGroup(DEFAULT_MODEL_GROUP)); // fill with default group
    return store;
}
