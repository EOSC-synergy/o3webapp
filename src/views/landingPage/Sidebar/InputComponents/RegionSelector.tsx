import React, { FC, Fragment } from 'react';
import { Box, Checkbox, FormControlLabel, Grid, Typography } from '@mui/material';
import { useDispatch, useSelector } from 'react-redux';
import {
    type RegionBasedXRange,
    selectPlotId,
    selectPlotXRange,
    setDisplayXRangeForPlot,
} from 'store/plotSlice';
import { ALL_REGIONS_ORDERED } from 'utils/constants';
import CustomLatitudeSelector from './LatitudeBandSelector/CustomLatitudeSelector';
import invariant from 'tiny-invariant';

/**
 * Enables the user to select / deselect regions as well as entering a custom region.
 *
 * @returns {JSX.Element}
 * @component
 * @see {@link LatitudeBandSelector}
 */
const RegionSelector: FC = () => {
    /** A dispatch function to dispatch actions to the redux store. */
    const dispatch = useDispatch();

    const activePlot = useSelector(selectPlotId);
    invariant(
        activePlot === 'tco3_return',
        'RegionSelector is only available for tco3_return plot'
    );

    const xRangeRegions = useSelector(selectPlotXRange(activePlot));

    const handleRegionChecked = (regionIdx: number) => {
        let regionCpy = [...xRangeRegions.regions];
        if (regionCpy.includes(regionIdx)) {
            regionCpy = regionCpy.filter((m) => m !== regionIdx);
        } else {
            regionCpy.push(regionIdx);
        }
        // Dispatch region checked
        regionCpy.sort();
        dispatch(
            setDisplayXRangeForPlot({
                plotId: activePlot,
                displayXRange: { regions: regionCpy },
            })
        );
    };

    return (
        <Grid container sx={{ width: '90%', marginLeft: 'auto', marginRight: 'auto' }}>
            <Typography>X-Axis:</Typography>
            <Box
                sx={{
                    paddingLeft: '8%',
                    paddingRight: '8%',
                    alignItems: 'left',
                    display: 'flex',
                    flexDirection: 'column',
                }}
            >
                {ALL_REGIONS_ORDERED.map((region, idx) => (
                    <Fragment key={idx}>
                        <FormControlLabel
                            label={
                                idx !== ALL_REGIONS_ORDERED.length - 1 ? (
                                    region
                                ) : (
                                    <CustomLatitudeSelector />
                                )
                            }
                            control={
                                <Checkbox
                                    checked={(xRangeRegions as RegionBasedXRange).regions.includes(
                                        idx
                                    )}
                                    onClick={() => handleRegionChecked(idx)}
                                    data-testid={`RegionSelector-${idx}`}
                                />
                            }
                        />
                    </Fragment>
                ))}
            </Box>
        </Grid>
    );
};

export default RegionSelector;
