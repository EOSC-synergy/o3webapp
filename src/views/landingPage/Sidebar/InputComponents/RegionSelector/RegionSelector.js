import React from "react";
import {Box, Checkbox, FormControlLabel, Typography} from "@mui/material";
import {useDispatch, useSelector} from "react-redux";
import {selectPlotXRange, setDisplayXRange} from "../../../../../store/plotSlice/plotSlice";
import {ALL_REGIONS_ORDERED} from "../../../../../utils/constants";
import {Grid} from "@mui/material";

/**
 * enables the user to select / deselect regions as well as entering a private region {@link LatitudeBandSelector}
 * 
 * @component
 * @param {Object} props
 * @param {function} props.reportError - used to report error functions
 * @returns {JSX.Element}
 */
function RegionSelector(props) {
    /**
     * A dispatch function to dispatch actions to the redux store.
     */
    const dispatch = useDispatch()

    /**
     * An array containing the selected regions.
     *
     * Examples:
     * If the first region is selected the array would have the following form: [0]
     * If the second and fifth region are selected the array would have the following form: [1, 4]
     */
    const xRangeRegions = useSelector(selectPlotXRange);
    /**
     * Handles the change if a region is clicked (selected/deselected).
     *
     * @param {number} regionIdx The index of the region that was clicked.
     */
    const handleRegionChecked = (regionIdx) => {
        let regionCpy = [...xRangeRegions.regions];
        if (regionCpy.includes(regionIdx)) {
            regionCpy = regionCpy.filter((m) => m !== regionIdx);
        } else {
            regionCpy.push(regionIdx);
        }
        // Dispatch region checked
        regionCpy.sort();
        dispatch(setDisplayXRange({regions: regionCpy})); // TODO
    }

    /**
     * Gets default regions that are available in the return recovery plot.
     * @todo connect to api
     */
    const getDefaultRegions = () => {
        return ALL_REGIONS_ORDERED;
    }

    return (
        <Grid container sx={{width: "90%", marginLeft: "auto", marginRight: "auto", marginTop: "3%"}}>
            <Typography style={{marginTop: '2.5%'}}>X-Axis:</Typography>
            <Box sx={{
                paddingLeft: '8%',
                paddingRight: '8%',
                alignItems: "left",
                display: "flex",
                flexDirection: "column"
            }}>
                {
                    getDefaultRegions().map((r, idx) => (
                        <React.Fragment key={idx}>
                            <FormControlLabel
                                label={r}
                                control={
                                    <Checkbox
                                        checked={xRangeRegions.regions.includes(idx)}
                                        onClick={() => handleRegionChecked(idx)}
                                    />
                                }
                            />
                        </React.Fragment>
                    ))
                }
            </Box>
        </Grid>
    );
}

export default RegionSelector;