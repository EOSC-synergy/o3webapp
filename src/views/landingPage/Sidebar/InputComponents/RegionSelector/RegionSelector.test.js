import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux'
import RegionSelector from './RegionSelector';
import { createTestStore } from "../../../../../store/store"
import {setActivePlotId} from "../../../../../store/plotSlice/plotSlice";

let store;
beforeEach(() => {
    store = createTestStore();
    store.dispatch(setActivePlotId({plotId: "tco3_return"})); // this component is only for the 
});

describe("tests basic rendering", () => {
    it('renders without crashing', () => {
        render(<>
            <Provider store={store}>
                <RegionSelector reportError={() => {}} />
            </Provider>
        </>)
    });
    
    it('renders correctly', () => {
    
        const { container } = render(<>
            <Provider store={store}>
                <RegionSelector reportError={() => {}} />
            </Provider>
        </>);
    
        expect(container).toMatchSnapshot();
    });
});
