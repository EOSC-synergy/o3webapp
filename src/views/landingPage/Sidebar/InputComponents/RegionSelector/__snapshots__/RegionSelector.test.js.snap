// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests basic rendering renders correctly 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container css-1ri4xmw-MuiGrid-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      style="margin-top: 2.5%;"
    >
      X-Axis:
    </p>
    <div
      class="MuiBox-root css-11wo1kh"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          Antarctic(Oct)
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          SH mid-lat
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          NH mid-lat
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          Tropics
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          Arctic(Mar)
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          Near global
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          Global
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          User region
        </span>
      </label>
    </div>
  </div>
</div>
`;
